### ISSUE TITLE
Client Side: Resolve Demarshall

### ISSUE THREAD

**Issue by thelimeburner** [gwAdvNet2015/gw-kv-store#50]
We plan to finish integration of the api into our code. Once the demarhsall client function is working we will integrate into our code and test it on a memcached server. This will complete the client.

**Comment by amygara** [comment]
Please write automated tests for your client and include them in your PR. I spoke with @jdk514 who said this should be doable.

**Comment by twood02** [comment]
what is the plan for automated tests for the client?  Is there any interesting way to test the client without having a server running somewhere?  I don't see how that is going to work with Travis...

**Comment by jdk514** [comment]
The tests would require that a server is running, but I think we should be
able to create tests that are independent of the server running. I imagine
we could simply send a "set" command, and then compare the value sent with
the response from a "get" request. Although this may not be what you were
originally thinking.

Joel

**Comment by amygara** [comment]
I think it may be easier to write tests if you try to emulate the memcached structure of doing things. By this, I mean create a library that will provide variables [that are provided in the standard libmemcached library](http://docs.libmemcached.org/libmemcached_examples.html) that allow the user to simply call an API that does all of this back-end work you've put in main and read_line. IE, have an API to add and connect to our gw-kv server, an API for SET (gwkv_set) and GET (gwkv_get). If you have this, you can simply have your testing platform be main and have it GET/SET and compare responses. Since we're designing a complex system, it would be best to have a good level of abstraction instead of leaving it up to the user to set everything up themselves, as you have now. This shouldn't be much work as all you should have to do is move your logic behind function calls and abstract it behind a header file. 

I'll be in the lab today from 2ish - 6ish, come see me and comment on this if you have other thoughts.

**Comment by twood02** [comment]
I think you guys should focus on cleaning up the code and separating it into functions.  Once we have that, we may see some easy tests we could put in place. For example, it would be good to have a test for your [read_line](https://github.com/sacul29/gw-kv-store/blob/apiclientversion/client/client-kv.c#L18) function.  If you change your `recv` calls to `read`, then you could have a test program that opened a file and passed the file descriptor to your function.  If you created the file containing the ASCII for a memcached protocol message, then you could verify that the struct you produce matches what you expect.  You might be able to do something similar for what your program needs to send (have it `write` instead of `send` and give it a file descriptor instead of a socket).  This way you can verify the client is producing the right sorts of messages, without having to worry about a server or libmemcached.

Basically I'm fine with the client tests being relatively simple as long as the code is well laid out.  I'd rather you do some manual tests with tools like valgrind to fix memory leaks, rather than build tests that won't be able to cover the important parts of the code anyway.

**Comment by twood02** [comment]
BTW I think you'll have to use raw `open` calls instead of the more common `fopen`.  See: http://codewiki.wikidot.com/c:system-calls:open for examples

**Comment by amygara** [comment]
Closed, for API specific stuff see #31. All marshal fixes have been completely merged into master.
